torch_geometric/explain/algorithm/gnn_explainer.py

```diff
@@ -0,0 +1,317 @@
+                             kwargs):
+        self._initialize_masks(x, edge_index,
+                               node_mask_type=explainer_config.node_mask_type,
+                               edge_mask_type=explainer_config.edge_mask_type)
+        self.to(x.device)
+
+        if explainer_config.edge_mask_type == MaskType.object:
+            set_masks(model, self.edge_mask, edge_index, apply_sigmoid=True)
+            parameters = [self.node_mask, self.edge_mask]
+        else:
+            parameters = [self.node_mask]
+        optimizer = torch.optim.Adam(parameters, lr=self.lr)
+
+        for _ in range(1, self.epochs + 1):
+            optimizer.zero_grad()
+            h = x * self.node_mask.sigmoid()
+            out = model(x=h, edge_index=edge_index, **kwargs)
+            loss_value = self.loss(
+                out, target, return_type=model_config.return_type,
+                target_idx=target_index, node_index=index,
+                edge_mask_type=explainer_config.edge_mask_type,
+                model_mode=model_config.mode)
+            loss_value.backward(retain_graph=True)
+            optimizer.step()
+
+    def _initialize_masks(self, x: Tensor, edge_index: Tensor,
+                          node_mask_type: MaskType, edge_mask_type: MaskType):
+        (N, F), E = x.size(), edge_index.size(1)
+        std = 0.1
+        if node_mask_type == MaskType.object:
+            self.node_mask = torch.nn.Parameter(torch.randn(N, 1) * std)
+        elif node_mask_type == MaskType.attributes:
+            if self.shared_mask:
+                self.node_mask = torch.nn.Parameter(torch.randn(1, F) * std)
+            else:
+                self.node_mask = torch.nn.Parameter(torch.randn(N, F) * std)
+
+        if edge_mask_type == MaskType.object:
+            std = torch.nn.init.calculate_gain('relu') * sqrt(2.0 / (2 * N))
+            self.edge_mask = Parameter(torch.randn(E) * std)
+
+    def loss_regression(
+        self,
+        y_hat: torch.Tensor,
+        y: torch.Tensor,
+        target_idx: Optional[int] = None,
+        node_index: Optional[int] = None,
+    ):
+        if target_idx is not None:
+            y_hat = y_hat[..., target_idx].unsqueeze(0)
+            y = y[..., target_idx].unsqueeze(0)
+
+        if node_index is not None and node_index >= 0:
+            loss_ = torch.cdist(y_hat[node_index], y[node_index])
+        else:
+            loss_ = torch.cdist(y_hat, y)
+
+        return loss_
+
+    def loss_classification(
+        self,
+        y_hat: torch.Tensor,
+        y: torch.Tensor,
+        return_type: ModelReturnType,
+        target_idx: Optional[int] = None,
+        node_index: Optional[int] = None,
+    ):
+        if target_idx is not None:
+            y_hat = y_hat[target_idx]
+            y = y[target_idx]
+
+        if return_type == ModelReturnType.probs:
+            y_hat = y_hat.log()
+        if return_type == ModelReturnType.raw:
+            y_hat = y_hat.log_softmax(dim=-1)
```

Comment:
Lets move this chunk of code to a method in `base.py`. Lots of explainers will have to do similar processing before computing loss, wdyt?
Another think you could do is as was done in `gnn_explainer` have `loss` in `base.py` do this processing and then call `_loss` which is implemented in the child class.

Reply:
Yes we can have a `get_prediction` method. Actually the `Explainer` class also has one, maybe we should have it as a utility function ?

Reply:
I'm not sure about forcing the preprocessing by the base class though, will all the explainer algorithms use log_probs for computing their classification loss ?

Reply:
Yeah thats a good point, lets have utility method in the base class instead that converts the probabilities to log_prob.

Reply:
It's done 👍🏽